### PR TITLE
PATCH add/remove/replace extensions using path to extension but not to exact attribute

### PR DESCRIPTION
**Pull request by jasonfagerberg-toast**
When you attempt to replace an entire extension schema (using EnterpriseSchema as an example here), scimple throws a Null Pointer exception

```
java.lang.NullPointerException
	at org.apache.directory.scim.core.repository.DefaultPatchHandler.apply(DefaultPatchHandler.java:135)
	at org.apache.directory.scim.core.repository.DefaultPatchHandler.apply(DefaultPatchHandler.java:118)
	at org.apache.directory.scim.core.repository.PatchHandlerTest.applyReplaceEntireEnterpriseExtension(PatchHandlerTest.java:327)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
```

This is because it is not treating it as an entire resource replacement but an exact path 

image

### CODE REVIEW

**Comment by bdemers** [comment]
@jasonfagerberg-toast I added a fix to this PR.
It's late, and there are a lot of edge cases in this `PatchHandler`, this is one more.

I think you are right, and this is valid patch request, although there are a quark with how the Attribute's are parsed (see comments)

I also added a test case for a "REMOVE" of a full Extension.  I'm not sure if that one is a valid 🤔 
(the containing extension object doesn't have a mutability flag, but it's attributes might, which _could_ make it an illegal operation, I'll have to check with the SCIM working group on that). As is I feel it's better to this code path defined and tested, so it's easier to make changes in the future.

Please take a quick pass through the code and let me know if you think something is missing or it's confusing.  I'll try to take another look when i'm more awake 😴

**Comment by jasonfagerberg-toast** [comment]
> @jasonfagerberg-toast I added a fix to this PR. It's late, and there are a lot of edge cases in this PatchHandler, this is one more.

Wow this is awesome! Wasn't expecting such a quick turn around! From my initial pass this looks good, I found another example online of a PATCH body doing the same thing in a slightly different way: https://is.docs.wso2.com/en/latest/references/extend/user-mgt/provisioning/extend-scim2-user-schemas/

```json
{
   "schemas":[
      "urn:ietf:params:scim:api:messages:2.0:PatchOp"
   ],
   "Operations":[
      {
         "op":"replace",
         "value":{
            "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User":{
               "customAttribute":"new value"
            }
         }
      }
   ]
}
```

So I added a test for that as well, and looks like it is passing 😄

File: scim-core/src/test/java/org/apache/directory/scim/core/repository/PatchHandlerTest.java [hunk 1]
```diff
@@ -56,6 +57,43 @@ public PatchHandlerTest() {
...
 
+  @Test
+  public void applyReplaceEntireEnterpriseExtension() {
+    String enterpriseExtensionUrn = "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User";
+    Map<String, String> enterpriseExtensionValue = new HashMap<>();
+    enterpriseExtensionValue.put("costCenter", "New Cost Center");
+    enterpriseExtensionValue.put("department", "New Department");
+    PatchOperation op = patchOperation(REPLACE, enterpriseExtensionUrn, enterpriseExtensionValue);
+    ScimUser updatedUser = patchHandler.apply(user(), List.of(op));
+    EnterpriseExtension actual = (EnterpriseExtension) updatedUser.getExtension("urn:ietf:params:scim:schemas:extension:enterprise:2.0:User");
```
Comment (jasonfagerberg-toast) [inline scim-core/src/test/java/org/apache/directory/scim/core/repository/PatchHandlerTest.java]: XS Nit:

This one is entirely my fault as I did this when I wrote the test, but we could replace the string here with the variable defined above
```suggestion
    EnterpriseExtension actual = (EnterpriseExtension) updatedUser.getExtension(enterpriseExtensionUrn);
```

File: scim-core/src/test/java/org/apache/directory/scim/core/repository/PatchHandlerTest.java [hunk 2]
```diff
@@ -56,6 +57,43 @@ public PatchHandlerTest() {
     this.patchHandler = new DefaultPatchHandler(schemaRegistry);
   }
 
+  @Test
+  public void applyReplaceEntireEnterpriseExtension() {
+    String enterpriseExtensionUrn = "urn:ietf:params:scim:schemas:extension:enterprise:2.0:User";
+    Map<String, String> enterpriseExtensionValue = new HashMap<>();
+    enterpriseExtensionValue.put("costCenter", "New Cost Center");
+    enterpriseExtensionValue.put("department", "New Department");
```
Comment (jasonfagerberg-toast) [inline scim-core/src/test/java/org/apache/directory/scim/core/repository/PatchHandlerTest.java]: XS Nit:

Again, entirely my fault, but from reading the rest of the tests, it looks like most use `Map.ofEntries` or `Map.of` to define an immutable map

File: scim-core/src/main/java/org/apache/directory/scim/core/repository/DefaultPatchHandler.java [hunk 3]
```diff
@@ -132,10 +132,23 @@ private <T extends ScimResource> void apply(T source, Map<String, Object> source
...
 
-      patchOperationHandler.applyExtensionValue(source, sourceAsMap, schema, attribute, valuePathExpression, attributeReference.getUrn(), patchOperation.getValue());
+      if (schema != null) {
+        Attribute attribute = schema.getAttribute(attributeReference.getAttributeName());
+        checkMutability(schema.getAttributeFromPath(attributeReference.getFullAttributeName()));
+
+        patchOperationHandler.applyExtensionValue(source, sourceAsMap, schema, attribute, valuePathExpression, attributeReference.getUrn(), patchOperation.getValue());
+      } else {
+        // If schema is null, it's either the root of an extension, or an invalid patch path
+        // It's not possible from the antlr parser to tell the diff between 'this:is:extension:urn' and 'this:is:extension:urn:attribute'
```
Comment (jasonfagerberg-toast) [inline scim-core/src/main/java/org/apache/directory/scim/core/repository/DefaultPatchHandler.java:143]: Yeah I was looking at this too, I think the SCIM [RFCs](https://datatracker.ietf.org/doc/html/rfc7644#section-3.4.2.2) ANBF has this same issue

```
attrPath  = [URI ":"] ATTRNAME *1subAttr
                 ; SCIM attribute name
                 ; URI is SCIM "schema" URI
``` 

its kinda unfortunate but I think this might be the path of least resistance and path of least hack

File: scim-core/src/test/java/org/apache/directory/scim/core/repository/PatchHandlerTest.java (hunk 1 above)
Comment (jasonfagerberg-toast) [reply scim-core/src/test/java/org/apache/directory/scim/core/repository/PatchHandlerTest.java]: Pushed this cleanup to the branch

File: scim-core/src/test/java/org/apache/directory/scim/core/repository/PatchHandlerTest.java (hunk 2 above)
Comment (jasonfagerberg-toast) [reply scim-core/src/test/java/org/apache/directory/scim/core/repository/PatchHandlerTest.java]: Pushed this cleanup to the branch

**Comment by jasonfagerberg-toast** [comment]
> Please take a quick pass through the code and let me know if you think something is missing or it's confusing. I'll try to take another look when i'm more awake 😴

I think for being sleepy this is an excellent PR 😅 I pushed a few small things to the tests but overall LGTM

**Comment by bdemers** [comment]
Thanks for taking a look at the updates!

File: scim-core/src/test/java/org/apache/directory/scim/core/repository/PatchHandlerTest.java [hunk 4]
```diff
@@ -19,6 +19,16 @@
 
 package org.apache.directory.scim.core.repository;
 
+import static java.util.Map.entry;
```
Comment (jasonfagerberg-toast) [inline scim-core/src/test/java/org/apache/directory/scim/core/repository/PatchHandlerTest.java]: My idea is probably set up with some linters that automatically re-ordered this and remove wildcard imports, I can undo it if you would like!

Comment (bdemers) [reply scim-core/src/test/java/org/apache/directory/scim/core/repository/PatchHandlerTest.java]: It's fine, it's isolated to a single test class so it's not the end of the world.
But if you want to clean it up or rewrite/squash any of the go for it.

If not, i'll try to merge this later today!

**Comment by bdemers** [comment]
Sorry about the delay on this one, I though we had already merged it!

Thanks again @jasonfagerberg-toast

**Comment by jasonfagerberg-toast** [comment]
> Thanks again @jasonfagerberg-toast 

Happy holidays :)

**Comment by jasonfagerberg-toast** [comment]
@bdemers just realized, this PR was closed and not merged, was that intentional? Mostly just asking when this change will be released
